src/Distribution/Server/Features/UserNotify.hs

```diff
@@ -797,18 +794,36 @@ userNotifyFeature serverEnv@ServerEnv{serverBaseURI, serverCron}
                     { notifyPackageInfo = pkg
                     }
 
-    genGroupUploadList notifyPrefs mp ga =
-        let (actor,gdesc) = case ga of (_,uid,Admin_GroupAddUser _ gd,_) -> (uid, gd)
-                                       (_,uid,Admin_GroupDelUser _ gd,_) -> (uid, gd)
-            addNotification uid m = if not (notifyOptOut npref) && notifyMaintainerGroup npref
-                                      then Map.insertWith (++) uid [ga] m
-                                      else m
-                where npref = fromMaybe defaultNotifyPrefs (Map.lookup uid notifyPrefs)
-        in case gdesc of
-           (MaintainerGroup pkg) -> do
-              maintainers <- queryUserGroup $ maintainersGroup (mkPackageName $ BS.unpack pkg)
-              return $ foldr addNotification mp (toList (delete actor maintainers))
-           _ -> return mp
+    genGroupUploadList notifyPrefs groupAction =
```

Comment:
All of these functions generating `Notification` values should probably be broken out + tested like `getUserNotificationsOnRelease`. But this PR is big enough already 🙃